### PR TITLE
Vendor 0.21.2

### PR DESCRIPTION
**Pull request by tcharding**
This is #77 without the version bump patch.

Upgrade to a new way of vendoring Core and vendor version 0.21.2 - note please this will be directly followed by an upgrade to 0.21-final. Done separately to proof out the workflow of such an upgrade.

### CODE REVIEW

**Comment by apoelstra** [comment]
I'm gonna go ahead and review/ACK this although strictly speaking the intermediate commits don't compile. I think this crate has historically had a lot of vendoring-related grief and bisection is probably already pretty busted, and this puts us on the start to fixing that.

**Comment by tcharding** [comment]
Yes, my bad I could have mentioned that - they explicitly do not compile, and I had exactly the same thoughts as you mention. It would be nice if we could start getting them all to compile from now. The do in secp so I think we should, I"ll try and keep that as a goal.

File: contrib/vendor-bitcoin-core.sh [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+DIR=./bitcoin
+
+pushd "$CORE_VENDOR_DEPEND_DIR" > /dev/null
+rm -rf "$DIR" || true
+
+# Clone the repo. As a special case, if the repo is a local path and we have
+# not specified a revision, just copy the directory rather than using 'git clone'.
+# This lets us use non-git repos or dirty source trees as secp sources.
+if [ "$CORE_VENDOR_CP_NOT_CLONE" == "yes" ]; then
+    cp -r "$CORE_VENDOR_REPO" .
```
Comment (apoelstra) [inline contrib/vendor-bitcoin-core.sh:75]: In b83e1debf3f2c24985ba37db0d29f326c21b179e:

In the rust-secp version of this it's a `cp "$VENDOR_REPO" "$DIR"`. Why did you change `$DIR` to `.`? It causes the following `chmod` to fail.

Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh:75]: At the time I couldn't work out a sane way to call the script, I was using 
```bash
 CORE_VENDOR_CP_NOT_CLONE=yes CORE_VENDOR_REPO=/home/tobin/build/github.com/tcharding/bitcoin contrib/vendor-bitcoin-core.sh v0.21-final
```

But that results in `bitcoin/bitcoin`?

Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh:75]: Are you using ` CORE_VENDOR_REPO=/path/to/bitcoin/*`

Comment (apoelstra) [reply contrib/vendor-bitcoin-core.sh:75]: No, just `CORE_VENDOR_REPO=/path/to/bitcoin`. As in

```
CORE_VENDOR_CP_NOT_CLONE=yes \
CORE_VENDOR_GIT_ROOT=.. \
CORE_VENDOR_REPO=/nix/store/77qhrhjq230jh6zswzg2nfxirpw4zdd3-source \
CORE_VENDOR_DEPEND_DIR=./depend2/ \
./contrib/vendor-bitcoin-core.sh -f
```

Comment (apoelstra) [reply contrib/vendor-bitcoin-core.sh:75]: If you are not specifying `CORE_VENDOR_DEPEND_DIR` then it will default to `depend/`. And you'll get a directory `depend/bitcoin` as expected. I'm not sure how you would get `bitcoin/bitcoin`.

Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh:75]: This is the verdor dir not the depend dir.

Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh:75]: This is the offending line
```
CORE_VENDOR_REPO=/nix/store/77qhrhjq230jh6zswzg2nfxirpw4zdd3-source
```
that directory is not called `bitcoin` like when I call it.

Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh:75]: I'll change it to how you use it, I can see how this is good for your usecase, when I do  `~/build/github.com` it doesn't seem good. But its a trifling matter.

Comment (apoelstra) [reply contrib/vendor-bitcoin-core.sh:75]: When you use `.` as the destination then it matters what the source directory is called because you are copying it into `.`. When you use a specific name, then you are copying the source directory to the new name (not *into* it) so it has exactly the same effect except that the source name doesn't matter. So I don't understand why this is worse for you.

Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh:75]: ooooo, TIL, if the destination directory exists already then the behaviour of `cp` is different to if it does not exist - I have never realized this, wow, mind blown. If the directory _does not exist_ it is as you say, it creates the directory correctly, but if it exists then it gets copied _into_ the existing one.

Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh:75]: I still don't know how this was happening though because we explicitly `rm -rf $DIR` right above. Anyways, onwards and upwards!

**Comment by tcharding** [comment]
Force pushed the change discussed above, and tested with `CORE_VENDOR_CP_NOT_CLONE=yes CORE_VENDOR_REPO=/home/tobin/build/github.com/tcharding/bitcoin contrib/vendor-bitcoin-core.sh v0.21.2`

Thanks for you patience :)

File: contrib/vendor-bitcoin-core.sh (hunk 1 above)
Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh]: > I have never realized this, wow, mind blown

FWIW I always use `rsync` and never use `cp -r`,  its still surprising that I did not know the nuances of `cp`.

Comment (apoelstra) [reply contrib/vendor-bitcoin-core.sh]: I wouldn't be *opposed* to using rsync here, FWIW, though it would be an extra dependency I'd have to pull into my nix derivation and it's probably overkill here.

Agreed that `cp` is nuts :)

File: contrib/vendor-bitcoin-core.sh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            CORE_VENDOR_VERSION="$1"
+        else
+            echo "WARNING: ignoring unknown command-line argument $1"
+        fi
+        ;;
+    esac
+    shift
+done
+
+echo "Vendoring Bitcoin Core version: $CORE_VENDOR_VERSION"
```
Comment (apoelstra) [inline contrib/vendor-bitcoin-core.sh:51]: In 902bc9ca085d9ebe2699044f369b225bcf0842e2:

The rust-secp version of this allows you to leave the revision unset and just use whatever code is present in the directory you're copying. We should preserve this behavior.

File: contrib/vendor-bitcoin-core.sh [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+if [ "$CORE_VENDOR_CP_NOT_CLONE" == "yes" ]; then
+    cp -r "$CORE_VENDOR_REPO" "$DIR"
+    chmod -R +w "$DIR" # cp preserves write perms, which if missing will cause patch to fail
+else
+    git clone "$CORE_VENDOR_REPO"
+fi
+
+# Check out specified revision
+pushd "$DIR" > /dev/null
+git checkout "$CORE_VENDOR_VERSION"
```
Comment (apoelstra) [inline contrib/vendor-bitcoin-core.sh:83]: In 902bc9ca085d9ebe2699044f369b225bcf0842e2:

In particular, here you unconditionally do a `git checkout`. This will not work if you are copying a source tree that is not in git.

File: contrib/vendor-bitcoin-core.sh (hunk 1 above)
Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh]: Oh I didn't mean to use it here, I just meant on my machine I never use `cp` for directories.

File: contrib/vendor-bitcoin-core.sh (hunk 2 above)
Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh:51]: Cracking the whip on me, I was too lazy to parse out the version number using regex - I'll do it now :(

File: contrib/vendor-bitcoin-core.sh (hunk 1 above)
Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh]: OMG, one last (hopefully) absolute face palm for the year
```
CORE_VENDOR_REPO=/home/tobin/build/github.com/tcharding/bitcoin
``` 

`github.com/bitcoin` is the org name, it contains the repostory `bitcoin` - it was not `cp`'s fault at all, it was all mine.

File: contrib/vendor-bitcoin-core.sh (hunk 2 above)
Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh]: Done, that wasn't so painful.

File: contrib/vendor-bitcoin-core.sh (hunk 3 above)
Comment (tcharding) [reply contrib/vendor-bitcoin-core.sh]: Now mimic's the secp script.

**Comment by tcharding** [comment]
Now more like the secp script and includes `-h` option.
```bash
$ contrib/vendor-bitcoin-core.sh -h

Usage: script OPTIONS [bitcoin-core-version]

OPTIONS:

 -f    Vendor even if there are local changes to the rust-bitcoinconsensus git index
 -h    Print this help an exit

Example:

    vendor-bitcoin-core v0.21.2
```

**Comment by apoelstra** [comment]
Awesome :) we are super close. The vendoring script works now, but when I check the vendored code against the source I see

```
diff -r depend/bitcoin/src/clientversion.cpp depend2/bitcoin/src/clientversion.cpp
28c28
< #define GIT_COMMIT_ID "12d7ec38e2c69539d6d9d7e99c4da8e41b7824d6"
---
> #define GIT_COMMIT_ID "af591f2068d0363c92d9756ca39c43db85e5804c"
```
These two commit IDs are the tip of this PR in this repo and the commit from Bitcoin Core that we're vendoring.

This is suuper weird. If you look at `bitcoin/src/clientversion.cpp` you will find neither line. Instead you will see
```
  //! git will put "#define GIT_COMMIT_ID ..." on the next line inside archives. $Format:%n#define GIT_COMMIT_ID "%H"$
```
How does `git` do this and under what circumstances? I don't know. But I worry that it is putting our commit ID into this line rather than the Core commit ID. OTOH I don't think any of these constants are used by libbitcoinconsensus so maybe it doesn't matter..

Anyway I think I should just whitelist this specific file and forget about it. But let me know what you think.

**Comment by tcharding** [comment]
> But I worry that it is putting our commit ID into this line rather than the Core commit ID.

I don't really like the idea of having this change in there, could bite us down the track. A simple hack could be, at the end of the vendoring script, to just grep for `GIT_COMMIT_ID` and echo a message "manually remove the change to GIT_COMMIT_ID".

**Comment by tcharding** [comment]
> #define GIT_COMMIT_ID "12d7ec38e2c69539d6d9d7e99c4da8e41b7824d6"

I had a bit of a play and we are still doing something different, I don't get these lines in `clientversion.cpp`, and they don't exist on this branch?

I tried using clone (and last time I pushed I was using copy)?

**Comment by apoelstra** [comment]
I think you need to use `git archive` or something to populate these lines. I am getting it by using Nix to source the git repo.

I think this is a non-issue because:
* The offending line doesn't actually appear in the repo; it is already dependent on some git magic that users are unlikely to trigger.
* The offending `#define` is not actually used in bitcoinconsensus anywhere.

**Comment by tcharding** [comment]
Cool, so this is good to go in then, right?

**Comment by apoelstra** [comment]
Yep. I'll tweak my scripts to ignore this line and ACK/merge it.

**Review by apoelstra — APPROVED** [review]
ACK 12d7ec38e2c69539d6d9d7e99c4da8e41b7824d6
